File: app/src/main/java/org/amoustakos/boilerplate/ui/activities/BaseActivity.java
package org.amoustakos.boilerplate.ui.activities;

import android.content.Context;
import android.os.Bundle;
import android.support.annotation.LayoutRes;
import android.support.annotation.Nullable;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.util.LongSparseArray;
import android.view.View;
import android.view.ViewGroup;

import org.amoustakos.boilerplate.BoilerplateApplication;
import org.amoustakos.boilerplate.R;
import org.amoustakos.boilerplate.injection.component.ActivityComponent;
import org.amoustakos.boilerplate.injection.component.ConfigPersistentComponent;
import org.amoustakos.boilerplate.injection.component.DaggerConfigPersistentComponent;
import org.amoustakos.boilerplate.injection.module.ActivityModule;

import java.util.concurrent.atomic.AtomicLong;

import butterknife.ButterKnife;
import timber.log.Timber;
import uk.co.chrisjenx.calligraphy.CalligraphyContextWrapper;

/**
 * Abstract activity that every other Activity in this application must implement. It handles
 * creation of Dagger components and makes sure that instances of ConfigPersistentComponent survive
 * across configuration changes.
 */
public abstract class BaseActivity extends AppCompatActivity {

    private static final String KEY_ACTIVITY_ID = "KEY_ACTIVITY_ID";
    private static final AtomicLong NEXT_ID = new AtomicLong(0);
	private static final LongSparseArray<ConfigPersistentComponent> sComponentsMap = new LongSparseArray<>();

    private ActivityComponent mActivityComponent;
    private long mActivityId;

    private ViewGroup rootView;


	/*
	 * Overridden
	 */

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
	    setContentView(layoutId());
	    makeID(savedInstanceState);
	    makeComponents(mActivityId);

	    Toolbar toolbar = getToolbar();
	    if (toolbar != null)
	    	setSupportActionBar(toolbar);
    }

	@Override
	protected void onSaveInstanceState(Bundle outState) {
		super.onSaveInstanceState(outState);
		outState.putLong(KEY_ACTIVITY_ID, mActivityId);
	}

	@Override
	protected void onDestroy() {
		if (!isChangingConfigurations()) {
			Timber.d("Clearing ConfigPersistentComponent id=%d", mActivityId);
			sComponentsMap.remove(mActivityId);
		}
		super.onDestroy();
	}

	@Override
	protected void attachBaseContext(Context newBase) {
		//Calligraphy
		super.attachBaseContext(CalligraphyContextWrapper.wrap(newBase));
	}


    /*
     * Injection
     */

    private void makeID(Bundle savedInstanceState) {
	    mActivityId = savedInstanceState != null ?
			    savedInstanceState.getLong(KEY_ACTIVITY_ID) : NEXT_ID.getAndIncrement();
    }

    private void makeComponents(long activityID) {
	    ConfigPersistentComponent configPersistentComponent;
	    int index = sComponentsMap.indexOfKey(activityID);
	    if (index < 0) { //component does not exist
		    Timber.d("Creating new ConfigPersistentComponent id=%d", activityID);
		    configPersistentComponent = DaggerConfigPersistentComponent.builder()
				    .applicationComponent(application().getComponent())
				    .build();
		    sComponentsMap.put(activityID, configPersistentComponent);
	    }
	    else {
		    Timber.d("Reusing ConfigPersistentComponent id=%d", activityID);
		    configPersistentComponent = sComponentsMap.get(index);
	    }
	    mActivityComponent = configPersistentComponent.activityComponent(new ActivityModule(this));
    }




    /*
     * View helpers
     */
    @LayoutRes protected abstract int layoutId();

    protected View getRootView() {
	    return ((ViewGroup)ButterKnife.findById(this, android.R.id.content)).getChildAt(0);
    }

    @Nullable protected Toolbar getToolbar() {
	    return (Toolbar) findViewById(R.id.toolbar);
    }




    /*
     * Getters
     */

    public ActivityComponent activityComponent() {
        return mActivityComponent;
    }

    public BoilerplateApplication application() {
	    return BoilerplateApplication.get(this);
    }

}

File: app/src/main/java/org/amoustakos/boilerplate/ui/activities/BaseActivity.kt
package org.amoustakos.boilerplate.ui.activities

import android.os.Bundle
import android.util.LongSparseArray
import androidx.annotation.LayoutRes
import androidx.appcompat.app.AppCompatActivity
import org.amoustakos.boilerplate.BoilerplateApplication
import org.amoustakos.boilerplate.di.component.ActivityComponent
import org.amoustakos.boilerplate.di.component.ConfigPersistentComponent
import org.amoustakos.boilerplate.di.component.DaggerConfigPersistentComponent
import org.amoustakos.boilerplate.di.module.injectors.ActivityModule
import org.amoustakos.boilerplate.view.base.ActivityViewComponent
import timber.log.Timber
import java.util.concurrent.atomic.AtomicLong

/**
 * Abstract activity that every other Activity in this application must implement. It handles
 * creation of Dagger components and makes sure that instances of ConfigPersistentComponent survive
 * across configuration changes.
 */
abstract class BaseActivity : AppCompatActivity() {

    private var mActivityComponent: ActivityComponent? = null
    private var mActivityId: Long = 0

//    protected val rootView: View
//        get() = (findViewById<View>(android.R.id.content) as ViewGroup).getChildAt(0)

    protected fun setupViewComponents(components: List<ActivityViewComponent>) {
	    components.forEach { setupViewComponent(it) }
    }

	protected fun setupViewComponent(component: ActivityViewComponent) {
		component.setup(this)
	}


    // =========================================================================================
    // Overridden
    // =========================================================================================

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(layoutId())
        makeID(savedInstanceState)
        makeComponent(mActivityId)
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putLong(KEY_ACTIVITY_ID, mActivityId)
    }

    override fun onDestroy() {
        if (!isChangingConfigurations) {
            Timber.d("Clearing ConfigPersistentComponent id=%d", mActivityId)
            sComponentsMap.remove(mActivityId)
        }
        super.onDestroy()
    }

    // =========================================================================================
    // Injection
    // =========================================================================================

    private fun makeID(savedInstanceState: Bundle?) {
        mActivityId = savedInstanceState?.getLong(KEY_ACTIVITY_ID) ?: NEXT_ID.getAndIncrement()
    }

    private fun makeComponent(activityID: Long) {
        val configPersistentComponent: ConfigPersistentComponent
        val index = sComponentsMap.indexOfKey(activityID)

        if (index < 0) { //component does not exist
            Timber.d("Creating new ConfigPersistentComponent id=%d", activityID)
            configPersistentComponent = DaggerConfigPersistentComponent.builder()
                    .applicationComponent(application().component)
                    .build()
            sComponentsMap.put(activityID, configPersistentComponent)
        } else {
            Timber.d("Reusing ConfigPersistentComponent id=%d", activityID)
            configPersistentComponent = sComponentsMap.get(index.toLong())
        }

        mActivityComponent = configPersistentComponent.activityComponent(ActivityModule(this))
    }


    // =========================================================================================
    // View helpers
    // =========================================================================================

    @LayoutRes
    protected abstract fun layoutId(): Int

    // =========================================================================================
    // Getters
    // =========================================================================================

    @Throws(IllegalStateException::class)
    fun activityComponent(): ActivityComponent {
        return mActivityComponent ?:
                throw IllegalStateException("Activity component not instantiated")
    }

    fun application() = BoilerplateApplication[this]

    companion object {
        private const val KEY_ACTIVITY_ID = "KEY_ACTIVITY_ID"
        private val NEXT_ID = AtomicLong(0)
        private val sComponentsMap = LongSparseArray<ConfigPersistentComponent>()
    }

}
